frontends/concrete-python/concrete/fhe/compilation/compiler.py

```diff
@@ -479,56 +487,123 @@ def compile(
+                if is_first:  # pragma: no cover
                     print()
+                    is_first = False
 
-                    print("Computation Graph")
-                    print("-" * columns)
-                    print(graph)
-                    print("-" * columns)
+                print("Bit-Width Constraints")
+                print("-" * columns)
+                for i, node in enumerate(nx.lexicographical_topological_sort(self.graph.graph)):
+                    if len(node.bit_width_constraints) > 0:
+                        print(f"%{i}:")
+                        for constraint in node.bit_width_constraints:
+                            print(
+                                f"    {constraint.arg(0)} {constraint.decl()} {constraint.arg(1)}"
+                            )
+                print("-" * columns)
+
+                print()
 
+            if show_bit_width_assignments:
+                if is_first:  # pragma: no cover
                     print()
+                    is_first = False
 
-                if show_mlir:
-                    print("\n" if not show_graph else "", end="")
+                print("Bit-Width Assignments")
+                print("-" * columns)
 
-                    print("MLIR")
-                    print("-" * columns)
-                    print(mlir_str)
-                    print("-" * columns)
+                lines = []
+                for variable in self.graph.bit_width_assignments.decls():  # type: ignore
+                    width = self.graph.bit_width_assignments.get_interp(variable)  # type: ignore
+                    lines.append(f"{variable} = {width}")
+
+                def sorter(line: str) -> int:
+                    if line.startswith("max"):
+                        # we won't have 4 million nodes...
+                        return 2**32
+
+                    assert line.startswith("%")
 
+                    equals_position = line.find("=")
+                    index = line[1 : equals_position - 1]
+                    return int(index)
+
+                longest_length_before_equals_sign = max(len(line.split("=")[0]) for line in lines)
+                for line in sorted(lines, key=sorter):
+                    length_before_equals_sign = len(line.split("=")[0])
+                    print(
+                        (" " * (longest_length_before_equals_sign - length_before_equals_sign))
+                        + line
+                    )
+
+                print("-" * columns)
+
+                print()
+
+            if show_assigned_graph:
+                if is_first:  # pragma: no cover
                     print()
+                    is_first = False
 
-                if show_optimizer:
-                    print("\n" if not (show_graph or show_mlir) else "", end="")
+                print("Bit-Width Assigned Computation Graph")
+                print("-" * columns)
+                print(self.graph.format(show_assigned_bit_widths=True))
```

Comment:
The graph is printed twice ? or is it a completely different format (non redundant) ?

Reply:
The same information is not printed twice. Both is necessary.